modules/javafx.graphics/src/main/java/javafx/scene/layout/HeaderBarBase.java

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package javafx.scene.layout;
+
+import com.sun.glass.ui.WindowOverlayMetrics;
+import com.sun.javafx.stage.StageHelper;
+import com.sun.javafx.tk.quantum.WindowStage;
+import javafx.beans.property.ReadOnlyDoubleProperty;
+import javafx.beans.property.ReadOnlyDoubleWrapper;
+import javafx.beans.property.ReadOnlyObjectWrapper;
+import javafx.geometry.Dimension2D;
+import javafx.geometry.HorizontalDirection;
+import javafx.scene.Node;
+import javafx.scene.Scene;
+import javafx.stage.Stage;
+import javafx.stage.StageStyle;
+import javafx.stage.Window;
+import javafx.util.Subscription;
+
+/**
+ * Base class for a client-area header bar that is used as a replacement for the system-provided header bar
+ * in stages with the {@link StageStyle#EXTENDED} style. This class is intended for application developers
+ * to use as a starting point for custom header bar implementations, and it enables the <em>click-and-drag</em>
+ * and <em>double-click to maximize</em> behaviors that are usually afforded by system-provided header bars.
+ * The entire {@code HeaderBarBase} background is draggable by default, but its content is not. Applications
+ * can specify draggable content nodes of the {@code HeaderBarBase} with the {@link #setDraggable} method.
+ *
+ * @apiNote Most application developers should use the {@link HeaderBar} implementation instead of
```

Comment:
thanks for explanation.  but does it really warrant a base class, couldn't the app either set different components inside the existing HB, or maybe set a different HB?

is there anything that HeaderBar does that may not be needed in a custom title bar?

it just feels that HBBase is not needed.

Reply:
We need to think of this in terms of the user model. `HeaderBar` gives rise to three named areas (leading, center, trailing) and comes with a certain layout behavior for each of those areas.

Sure, it is possible to place another layout container in the `HeaderBar.center` area and set its prefWidth to infinity in order to effectively get rid of all of the custom layout behavior that `HeaderBar` brings to the table.

But that severely limits application and library developers from customizing the user model of the header bar. For example, using `HeaderBarBase`, a library developer can create alternative header bars with a different set of features and a different user model. Those alternative header bars might have different layout areas, different properties, in short a different API. It would be really awkward to shoehorn all of that into the (opinionated) API of `HeaderBar`.

Reply:
I understand that, but could you give a **specific** example?

There is nothing in HB that prevents it from being used in all the scenarios I can think of.  It's like BorderPane - it can be used with a single center component, top and center, etc.  I just don't think we need a separate base class in this particular case, unless something in the HB **_prevents_** it.

Reply:
`BorderPane` is a good example. It's a customized API on top of `Pane`, which is basically just defined as a "collection of children":
> Base class for layout panes which need to expose the children list as public
> so that users of the subclass can freely add/remove children.

You're arguing that you wouldn't need `Pane` as a base class, because you can just as easily build a `VBox` on top of a `BorderPane` if you only use the center area and ignore all of the rest. While that's true, it's not a good user model.

I already gave you a specific example (Spotify) where an alternative implementation would dynamically change alignments, and presumably have a different API because of that.

As another example, take a look at what our WinUI friends are doing with their [TitleBar](https://github.com/microsoft/microsoft-ui-xaml/issues/9700). They have more API elements than our implementation (a back button, pane toggle button, icon, title, subtitle, and so on). The point of this example is _not_ to argue whether JavaFX developers could recreate this specific API by subclassing `HeaderBar`, it is to argue that custom APIs are relevant, and only `HeaderBarBase` gives developers a clean slate to create their own API.

Reply:
Perhaps I am not entirely clear.  But having these examples helped, thanks.

My point is that, in my opinion, we don't need two classes.  We only need one, with a single `'center'` property.  If the app needs leading/tailing/upper/lower/anythingElse, the app is free to use any container suitable for their need.

This way the platform provides the basis, which only the platform can provide.

So my suggestion would be to combine two classes into one with a single `'center'` property.

What do you think?

Reply:
I see where you're coming from. While that might be a possible way to skin the cat, I think it has a few downsides. It would remove most API from `HeaderBar`, and instead, as you say, accept a single node as content. This would shift the concrete API from `HeaderBar` to whatever node/pane/layout container is placed in it.

However, the layout container inside this hypothetical header bar would, for many scenarios, need to have back-references to the header bar in order to achieve its purpose, either directly or indirectly by inspecting its own parents. One example is the centering behavior currently implemented for `HeaderBar.center` with a center-aligned node: it needs to know how wide the system insets are to compute the layout. This would mean that a layout container needs to be created that is aware of the context in which it is used, and changes its behavior (or maybe only meaningfully works) if it is placed in a `HeaderBar`.

So even recreating the current `HeaderBar` would lead to a bad and potentially fragile API. This, in my opinion, rules out the approach to shifting all layout behavior to a user-supplied layout container.

Reply:
aha!  this is the (good) argument I was asking for, thanks!

I mean, I am not buying it entirely, since "centering" may mean different things to different people - for example using window half-width may or may not be a good centering position if you have asymmetry due to the buttons on one side, perhaps it would be better to center across the empty area instead.

But if other people are ok with breaking up this thing into two classes, I won't object (much).